File: src/main/java/dev/lbuddyboy/flash/rank/impl/MongoRank.java
package dev.lbuddyboy.flash.rank.impl;

import com.mongodb.client.model.Filters;
import com.mongodb.client.model.ReplaceOptions;
import dev.lbuddyboy.flash.Flash;
import dev.lbuddyboy.flash.rank.Rank;
import dev.lbuddyboy.flash.rank.packet.RanksUpdatePacket;
import dev.lbuddyboy.flash.util.Tasks;
import dev.lbuddyboy.flash.util.gson.GSONUtils;
import org.bson.Document;
import org.bukkit.ChatColor;

import java.util.UUID;

public class MongoRank extends Rank {

    public MongoRank(String name) {
        this.uuid = UUID.randomUUID();
        this.name = name;

        load();
    }

    public MongoRank(UUID uuid, String name) {
        this.uuid = uuid;
        this.name = name;

        load();
    }

    @Override
    public void load() {
        Document document = Flash.getInstance().getMongoHandler().getRankCollection().find(Filters.eq("uuid", this.uuid.toString())).first();

        if (document == null) {
            save(true, true);
            return;
        }

        if (document.containsKey("name"))
            this.name = document.getString("name");

        if (document.containsKey("displayName"))
            this.displayName = document.getString("displayName");

        if (document.containsKey("color"))
            this.color = ChatColor.valueOf(document.getString("color"));

        if (document.containsKey("weight"))
            this.weight = document.getInteger("weight");

        if (document.containsKey("prefix"))
            this.prefix = document.getString("prefix");

        if (document.containsKey("suffix"))
            this.suffix = document.getString("suffix");

        if (document.containsKey("default"))
            this.defaultRank = document.getBoolean("default");

        if (document.containsKey("permissions"))
            this.permissions = GSONUtils.getGSON().fromJson(document.getString("permissions"), GSONUtils.STRING);

        if (document.containsKey("inheritance"))
            this.inheritance = GSONUtils.getGSON().fromJson(document.getString("inheritance"), GSONUtils.STRING);

    }

    @Override
    public void delete() {
        Flash.getInstance().getRankHandler().getRanks().remove(this.uuid);
        Flash.getInstance().getMongoHandler().getRankCollection().deleteOne(Filters.eq("uuid", this.uuid.toString()));
        new RanksUpdatePacket(Flash.getInstance().getRankHandler().getRanks()).send();
    }

    @Override
    public void save(boolean async) {
        save(async, false);
    }

    private void save(boolean async, boolean reload) {
        if (!async) {
            Document document = new Document();

            document.put("uuid", this.uuid.toString());
            document.put("name", this.name);
            document.put("displayName", this.displayName);
            document.put("color", this.color.name());
            document.put("weight", this.weight);
            document.put("prefix", this.prefix);
            document.put("suffix", this.suffix);
            document.put("permissions", GSONUtils.getGSON().toJson(this.permissions, GSONUtils.STRING));

            Flash.getInstance().getMongoHandler().getRankCollection().replaceOne(Filters.eq("uuid", this.uuid.toString()), document, new ReplaceOptions().upsert(true));

            if (reload) load();
            return;
        }
        Tasks.runAsync(() -> save(false, reload));
    }

}

File: src/main/java/dev/lbuddyboy/flash/rank/impl/MongoRank.kt
package dev.lbuddyboy.flash.rank.impl

import com.mongodb.client.model.Filtersimport

com.mongodb.client.model.ReplaceOptionsimport dev.lbuddyboy.flash.Flashimport dev.lbuddyboy.flash.rank.Rankimport dev.lbuddyboy.flash.rank.packet.RanksUpdatePacketimport dev.lbuddyboy.flash.util.gson.GSONUtilsimport org.bson.Documentimport org.bukkit.*import java.util.*

class MongoRank : Rank {
    constructor(name: String?) {
        uuid = UUID.randomUUID()
        this.name = name
        load()
    }

    constructor(uuid: UUID?, name: String?) {
        this.uuid = uuid
        this.name = name
        load()
    }

    override fun load() {
        val document: Document =
            Flash.instance.mongoHandler.getRankCollection().find(Filters.eq("uuid", uuid.toString())).first()
        if (document == null) {
            save(true, true)
            return
        }
        if (document.containsKey("name")) name = document.getString("name")
        if (document.containsKey("displayName")) displayName = document.getString("displayName")
        if (document.containsKey("color")) color = ChatColor.valueOf(document.getString("color"))
        if (document.containsKey("weight")) weight = document.getInteger("weight")
        if (document.containsKey("prefix")) prefix = document.getString("prefix")
        if (document.containsKey("suffix")) suffix = document.getString("suffix")
        if (document.containsKey("default")) defaultRank = document.getBoolean("default")
        if (document.containsKey("staff")) staff = document.getBoolean("staff")
        if (document.containsKey("permissions")) permissions =
            GSONUtils.getGSON().fromJson(document.getString("permissions"), GSONUtils.STRING)
        if (document.containsKey("inheritance")) inheritance =
            GSONUtils.getGSON().fromJson(document.getString("inheritance"), GSONUtils.STRING)
    }

    override fun delete() {
        Flash.instance.rankHandler.getRanks().remove(uuid)
        Flash.instance.mongoHandler.getRankCollection().deleteOne(Filters.eq("uuid", uuid.toString()))
        RanksUpdatePacket(Flash.instance.rankHandler.getRanks()).send()
    }

    override fun save(async: Boolean) {
        save(async, false)
    }

    private fun save(async: Boolean, reload: Boolean) {
        val document = Document()
        document["uuid"] = uuid.toString()
        document["name"] = name
        document["displayName"] = displayName
        document["color"] = color.name
        document["weight"] = weight
        document["prefix"] = prefix
        document["suffix"] = suffix
        document["default"] = defaultRank
        document["staff"] = staff
        document["permissions"] = GSONUtils.getGSON().toJson(permissions, GSONUtils.STRING)
        document["inheritance"] = GSONUtils.getGSON().toJson(inheritance, GSONUtils.STRING)
        Flash.instance.mongoHandler.getRankCollection()
            .replaceOne(Filters.eq("uuid", uuid.toString()), document, ReplaceOptions().upsert(true))
        if (reload) load()
    }
}